File: leetcode/easy/217.contains-duplicate/solutions/go/solution1.go
/**
 * Solution 1: Map-based Approach (Equivalent to HashSet)
 *
 * This approach uses a Go map to track elements we've seen so far.
 * If we encounter an element that's already in the map, we found a duplicate.
 *
 * Time Complexity: O(n) - Single pass through the array
 * Space Complexity: O(n) - Map can contain up to n elements
 *
 * Advantages:
 * - Optimal time complexity for this problem
 * - Simple and intuitive
 * - Early termination when duplicate is found
 *
 * Disadvantages:
 * - Uses extra space proportional to input size
 * - Not suitable for space-constrained environments
 */
package main

// containsDuplicate checks if array contains any duplicates using a map
func containsDuplicate(nums []int) bool {
	// Early return for edge case
	if len(nums) <= 1 {
		return false
	}

	// Map to track seen elements
	seen := make(map[int]bool)

	// Iterate through each element
	for _, num := range nums {
		// If element already exists in map, we found a duplicate
		if seen[num] {
			return true // Early termination - duplicate found
		}
		// Add element to map
		seen[num] = true
	}

	// No duplicates found
	return false
}

// containsDuplicateStruct uses struct{} for better memory efficiency
// struct{} takes 0 bytes vs bool which takes 1 byte
func containsDuplicateStruct(nums []int) bool {
	if len(nums) <= 1 {
		return false
	}

	// Map with struct{} values for memory efficiency
	seen := make(map[int]struct{})

	for _, num := range nums {
		if _, exists := seen[num]; exists {
			return true // Duplicate found
		}
		seen[num] = struct{}{} // Add to map
	}

	return false
}

// containsDuplicateWithCapacity pre-allocates map capacity for better performance
func containsDuplicateWithCapacity(nums []int) bool {
	if len(nums) <= 1 {
		return false
	}

	// Pre-allocate map capacity for better performance
	seen := make(map[int]bool, len(nums))

	for _, num := range nums {
		if seen[num] {
			return true
		}
		seen[num] = true
	}

	return false
}

File: leetcode/easy/217.contains-duplicate/solutions/java/solution1.java
/**
 * Solution 1: HashSet Approach (Optimal)
 * 
 * This approach uses a HashSet to track elements we've seen so far.
 * If we encounter an element that's already in the set, we found a duplicate.
 * 
 * Time Complexity: O(n) - Single pass through the array
 * Space Complexity: O(n) - HashSet can contain up to n elements
 * 
 * Advantages:
 * - Optimal time complexity for this problem
 * - Simple and intuitive
 * - Early termination when duplicate is found
 * 
 * Disadvantages:
 * - Uses extra space proportional to input size
 * - Not suitable for space-constrained environments
 */
public class solution1 {
    
    /**
     * Check if array contains any duplicates using HashSet
     * 
     * @param nums Input array of integers
     * @return true if duplicates exist, false otherwise
     */
    public static boolean containsDuplicate(int[] nums) {
        // Early return for edge case
        if (nums == null || nums.length <= 1) {
            return false;
        }
        
        // HashSet to track seen elements
        java.util.Set<Integer> seen = new java.util.HashSet<>();
        
        // Iterate through each element
        for (int num : nums) {
            // If add() returns false, element already exists (duplicate found)
            if (!seen.add(num)) {
                return true; // Early termination - duplicate found
            }
        }
        
        // No duplicates found
        return false;
    }
    
    /**
     * Alternative implementation using contains() method
     * This is slightly less efficient but more explicit
     */
    public static boolean containsDuplicateAlternative(int[] nums) {
        if (nums == null || nums.length <= 1) {
            return false;
        }
        
        java.util.Set<Integer> seen = new java.util.HashSet<>();
        
        for (int num : nums) {
            if (seen.contains(num)) {
                return true; // Duplicate found
            }
            seen.add(num); // Add to set
        }
        
        return false;
    }
}
